cannon/mipsevm/tests/difftester_test.go

```diff
@@ -0,0 +1,364 @@
+				state.GetRegistersRef()[2] = syscallNum
+			}
+
+			expectationsCalled := make(map[string]int)
+			setExpectations := func(testCase simpleTestCase, expect *mtutil.ExpectedState, vm VersionedVMTestCase) ExpectedExecResult {
+				expectationsCalled[testCase.name] += 1
+				expect.ExpectStep()
+
+				if useCorrectReturnExpectation {
+					return ExpectPanic("unrecognized syscall: 0", "unimplemented syscall")
+				} else {
+					return ExpectNormalExecution()
+				}
+			}
+
+			versions := GetMipsVersionTestCases(t)
+			expectedTestCases := generateExpectedTestCases(testCases, versions)
+
+			// Run tests
+			tRunner := newMockTestRunner(t)
+			NewDiffTester(testNamer).
+				InitState(initState).
+				SetExpectations(setExpectations).
+				run(tRunner, testCases, 0)
+
+			// Validate that we invoked initState and setExpectations as expected
+			for _, c := range testCases {
+				testsPerCase := len(versions)
+				require.Equal(t, testsPerCase, initStateCalled[c.name])
+				// Difftester runs extra calls on the expectations fn in order to analyze the tests
+				require.Equal(t, testsPerCase+len(versions), expectationsCalled[c.name])
+			}
+
+			// Validate that tests ran and passed as expected
+			require.Equal(t, len(tRunner.childTestMocks), len(expectedTestCases))
+			for _, testCase := range expectedTestCases {
+				if useCorrectReturnExpectation {
+					failed, err := tRunner.testFailedOrPanicked(testCase)
+					require.NoError(t, err)
+					require.False(t, failed, "Test '%v' should pass", testCase)
+				} else {
+					panicked, err := tRunner.testPanicked(testCase)
+					require.NoError(t, err)
+					require.True(t, panicked, "Test '%v' should panic", testCase)
+				}
+			}
+		})
+	}
+}
+
+// Test case struct for simple test scenarios
+type simpleTestCase struct {
+	name string
+	insn uint32
+}
+
+// Test helper to create a test namer
+func testNamer(testCase simpleTestCase) string {
+	return testCase.name
+}
+
+// generateExpectedTestCases Generates expected test cases that are derived from the provided `cases`
+func generateExpectedTestCases(cases []simpleTestCase, versions []VersionedVMTestCase, expectedMods ...string) []string {
+	var expectedTestRuns []string
+	expectedMods = append(expectedMods, "")
+	for _, vm := range versions {
+		for _, testCase := range cases {
+			for _, mod := range expectedMods {
+				testName := fmt.Sprintf("%v%v (%v)", testCase.name, mod, vm.Name)
+				expectedTestRuns = append(expectedTestRuns, testName)
+			}
+		}
+	}
+	return expectedTestRuns
+}
+
+type mockTestRunner struct {
+	*mockT
+	childTestMocks map[string]*mockT
+}
+
+var _ testRunner = (*mockTestRunner)(nil)
+
+func newMockTestRunner(tb testing.TB) *mockTestRunner {
+	t := &mockT{name: "MockTestRunner", t: tb, debugTestFailures: false}
+	childMocks := make(map[string]*mockT)
+	return &mockTestRunner{mockT: t, childTestMocks: childMocks}
+}
+
+func (m *mockTestRunner) Run(name string, fn testFn) bool {
+	// TODO: How should we handle panics?
```

Comment:
Since panics in the VM are explicitly caught by tests, a test shouldn't panic. Else it's a buggy test and I'd like that to be very visible with a stack trace, rather than result in a "failed" test.